Let torch.cpu.current_device() return the device index

egienvalue opened the pull request:
Summary: Change the torch.cpu.current_device() behavior to return device index instead of string. (align with torch.cuda.current_device()).

Test Plan: OSS CIs.

Differential Revision: D60803063


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

wz337 commented:
Could I ask the motivation for this change? 

Hmm. Upon checking the history of the change, this might break the current usage. For example: https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/optimizer.py#L99

egienvalue commented:
> Could I ask the motivation for this change?
> 
> Hmm. Upon checking the history of the change, this might break the current usage. For example: https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/optimizer.py#L99

Yea, we are trying to make high level application can use device agnostic APIs to interact with different device backends. It requires torch.get_device_module().current_device() has similar behavior on all device backends.

wz337 reviewed (approved):
Thanks for the context! LGTM!

egienvalue committed:
Let torch.cpu.current_device() return the device index (#132760)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/132760

Change the torch.cpu.current_device() behavior to return device index instead of string. (align with torch.cuda.current_device()).

Test Plan: OSS CIs.

Reviewed By: wz337, scottxu0730, nautsimon

Differential Revision: D60803063

kwen2501 commented on torch/cpu/__init__.py:
```diff
@@ -171,9 +171,9 @@ def set_device(device: _device_t) -> None:
 
-def current_device() -> str:
-    r"""Returns current device for cpu. Always 'cpu'.
+def current_device() -> int:
+    r"""Returns current device for cpu. Always 0.
```
I wonder if this would have backward compatibility issue.
Cc @albanD @janeyx99 to comment

janeyx99 replied:
Yea this is 100% not backwards compatible!! We should not do this without the proper deprecation cycle, and before then, the reason that this then "lines up with cuda" is not sufficient to go through with this.

janeyx99 reviewed (changes requested):
This change as is is BC-breaking
